File: cplus/leetcode/202606/algoiv04.01.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool findWhetherExistsPath(int n, vector<vector<int>>& graph, int start, int target) {
        vector<vector<int>> g(n);
        for (auto e: graph) {
            g[e[0]].push_back(e[1]);
        }

        vector<int> visited(n);

        auto dfs = [&](this auto&& dfs, int u) -> bool {
            if (u == target) return true;
            for (auto v: g[u]) {
                if (!visited[v]) {
                    visited[v] = true;
                    if (dfs(v)) return true;
                }
            }
            return false;
        };

        return dfs(start);
    }
};

File: python/algo/202606/algoiv04.01.py
from typing import List


class Solution:
    def findWhetherExistsPath(self, n: int, graph: List[List[int]], start: int, target: int) -> bool:
        g = [[] for _ in range(n)]
        for e in graph:
            g[e[0]].append(e[1])
        
        visited = [False] * n
        def dfs(u: int) -> bool:
            if u == target: return True

            for v in g[u]:
                if not visited[v]:
                    visited[v] = True
                    if dfs(v): return True
            
            return False

        return dfs(start)